### ISSUE TITLE
"make prepare" on Ubuntu 16.04 fails to unpack android-2.3.3_r02.zip

### ISSUE THREAD

**Issue by janusw** [xamarin/xamarin-android#4082]
### Steps to Reproduce

1. git clone https://github.com/xamarin/xamarin-android.git
2. cd xamarin-android
3. make prepare

### Expected Behavior

`make prepare` should run through without errors.

### Actual Behavior

`make prepare` successfully checks for all necessary programs and tools, but then fails with the following error while preparing the Android SDK:

```
  🔗 https://dl.google.com/android/repository/android-2.3.3_r02.zip
    unpacking android-2.3.3_r02.zip to /home/janus/android-toolchain/sdk/platforms/android-10
Archive path: /home/janus/android-archives/android-2.3.3_r02.zip
[7zip] extracting archive
[7zip] log file: bin/BuildDebug/prepare-20200105T135338.7zip-extract.android-2.3.3_r02.zip.log
Elapsed: 00:00:00.0346794

Step Xamarin.Android.Prepare.Step_Android_SDK_NDK failed: Failed to unpack /home/janus/android-archives/android-2.3.3_r02.zip
System.InvalidOperationException: Step Xamarin.Android.Prepare.Step_Android_SDK_NDK failed: Failed to unpack /home/janus/android-archives/android-2.3.3_r02.zip ---> System.InvalidOperationException: Failed to unpack /home/janus/android-archives/android-2.3.3_r02.zip
  at Xamarin.Android.Prepare.Step_Android_SDK_NDK.Unpack (Xamarin.Android.Prepare.Context context, System.String tempDirRoot, Xamarin.Android.Prepare.Step_Android_SDK_NDK+AndroidPackage pkg) [0x0013d] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Steps/Step_Android_SDK_NDK.cs:163 
  at Xamarin.Android.Prepare.Step_Android_SDK_NDK.Execute (Xamarin.Android.Prepare.Context context) [0x00686] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Steps/Step_Android_SDK_NDK.cs:87 
  at Xamarin.Android.Prepare.Step.Run (Xamarin.Android.Prepare.Context context) [0x0003e] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Application/Step.cs:42 
  at Xamarin.Android.Prepare.Scenario.Run (Xamarin.Android.Prepare.Context context, Xamarin.Android.Prepare.Log log) [0x000a9] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Application/Scenario.cs:33 
   --- End of inner exception stack trace ---
  at Xamarin.Android.Prepare.Scenario.Run (Xamarin.Android.Prepare.Context context, Xamarin.Android.Prepare.Log log) [0x00172] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Application/Scenario.cs:44 
  at Xamarin.Android.Prepare.Context.Execute () [0x000fc] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Application/Context.cs:790 
  at Xamarin.Android.Prepare.App.Run (System.String[] args) [0x00858] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Main.cs:164 
```

### Version Information

* xamarin-android version: master as of 2020-01-05 (commit afb389cbb3ea59d4254038e0f739052e4d374c40)
* OS: Ubuntu Linux 16.04 (xenial)
* Mono version: 6.4.0.198
* msbuild version: Microsoft (R)-Build-Engine, Version 16.3.0-ci für Mono
* 7z version: p7zip 9.20

### Log File

bin/BuildDebug/prepare-20200105T135338.7zip-extract.android-2.3.3_r02.zip.log:

```
Error:
Incorrect command line
```

bin/BuildDebug/prepare-20200105T135338.log:
```
...
[00:00:03.2086035] Android component 'android-2.3.3_r02' will be downloaded from https://dl.google.com/android/repository/android-2.3.3_r02.zip
[00:00:03.5518798]   🔗 https://dl.google.com/android/repository/android-2.3.3_r02.zip
[00:00:03.5539736] android-2.3.3_r02 URL: https://dl.google.com/android/repository/android-2.3.3_r02.zip
[00:00:03.5539816] About to start downloading
[00:00:03.5551885] Calling GetAsync
[00:00:03.6623875] GetAsync finished
[00:00:03.6648470] Downloading...
[00:00:04.6700372] 26,55% (21,6 MB at 21,6 MB/s)             [00:00:05.6708699] 56,72% (46,2 MB at 24,6 MB/s)             [00:00:06.6719834] 84,65% (69,0 MB at 22,8 MB/s)             [00:00:07.1884670] success == True[00:00:07.1895620]     unpacking android-2.3.3_r02.zip to [00:00:07.1895844] /home/janus/android-toolchain/sdk/platforms/android-10
[00:00:07.1896018] 7z binary path: /usr/bin/7za
[00:00:07.1941155] Creating directory: /home/janus/android-toolchain/temp/ng4om55s.6gs
[00:00:07.1941916] Unpacking /home/janus/android-archives/android-2.3.3_r02.zip to directory: /home/janus/android-toolchain/temp/ng4om55s.6gs
[00:00:07.1943390] Custom NuGet path: /usr/bin/7za
[00:00:07.1943737] 7zip executable path is rooted, using verbatim
[00:00:07.1943770] Full 7zip executable path value: /usr/bin/7za
[00:00:07.1953547] Archive path: /home/janus/android-archives/android-2.3.3_r02.zip
[00:00:07.1956101] [7zip] extracting archive
[00:00:07.1966954] [7zip] log file: [00:00:07.1967128] bin/BuildDebug/prepare-20200105T135338.7zip-extract.android-2.3.3_r02.zip.log
[00:00:07.1971499] Running: /usr/bin/7za "x" -bd -bso1 -bsp1 -bse2 -y "-o/home/janus/android-toolchain/temp/ng4om55s.6gs" "/home/janus/android-archives/android-2.3.3_r02.zip"
[00:00:07.2318590] Elapsed: 00:00:00.0346794[00:00:07.2318870] 
[00:00:07.2322605] Deleting file: /home/janus/android-archives/android-2.3.3_r02.zip
[00:00:07.2461138] 
[00:00:07.2461400] Step Xamarin.Android.Prepare.Step_Android_SDK_NDK failed: Failed to unpack /home/janus/android-archives/android-2.3.3_r02.zip
[00:00:07.2481015] System.InvalidOperationException: Step Xamarin.Android.Prepare.Step_Android_SDK_NDK failed: Failed to unpack /home/janus/android-archives/android-2.3.3_r02.zip ---> System.InvalidOperationException: Failed to unpack /home/janus/android-archives/android-2.3.3_r02.zip
[00:00:07.2481015]   at Xamarin.Android.Prepare.Step_Android_SDK_NDK.Unpack (Xamarin.Android.Prepare.Context context, System.String tempDirRoot, Xamarin.Android.Prepare.Step_Android_SDK_NDK+AndroidPackage pkg) [0x0013d] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Steps/Step_Android_SDK_NDK.cs:163 
[00:00:07.2481015]   at Xamarin.Android.Prepare.Step_Android_SDK_NDK.Execute (Xamarin.Android.Prepare.Context context) [0x00686] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Steps/Step_Android_SDK_NDK.cs:87 
[00:00:07.2481015]   at Xamarin.Android.Prepare.Step.Run (Xamarin.Android.Prepare.Context context) [0x0003e] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Application/Step.cs:42 
[00:00:07.2481015]   at Xamarin.Android.Prepare.Scenario.Run (Xamarin.Android.Prepare.Context context, Xamarin.Android.Prepare.Log log) [0x000a9] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Application/Scenario.cs:33 
[00:00:07.2481015]    --- End of inner exception stack trace ---
[00:00:07.2481015]   at Xamarin.Android.Prepare.Scenario.Run (Xamarin.Android.Prepare.Context context, Xamarin.Android.Prepare.Log log) [0x00172] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Application/Scenario.cs:44 
[00:00:07.2481015]   at Xamarin.Android.Prepare.Context.Execute () [0x000fc] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Application/Context.cs:790 
[00:00:07.2481015]   at Xamarin.Android.Prepare.App.Run (System.String[] args) [0x00858] in /home/janus/github/xamarin-android/build-tools/xaprepare/xaprepare/Main.cs:164 
[00:00:07.2481543] 
```

The above shows that the relevant 7z command line, which seems to be failing, is:

```
/usr/bin/7za "x" -bd -bso1 -bsp1 -bse2 -y "-o/home/janus/android-toolchain/temp/ng4om55s.6gs" "/home/janus/android-archives/android-2.3.3_r02.zip"
```

**Comment by janusw** [comment]
> The above shows that the relevant 7z command line, which seems to be failing, is:
> 
> ```
> /usr/bin/7za "x" -bd -bso1 -bsp1 -bse2 -y "-o/home/janus/android-toolchain/temp/ng4om55s.6gs" "/home/janus/android-archives/android-2.3.3_r02.zip"
> ```

The 7z help text on my system shows:

```
$ 7za

7-Zip (A) [64] 9.20  Copyright (c) 1999-2010 Igor Pavlov  2010-11-18
p7zip Version 9.20 (locale=de_DE.UTF-8,Utf16=on,HugeFiles=on,4 CPUs)

Usage: 7za <command> [<switches>...] <archive_name> [<file_names>...]
       [<@listfiles...>]

<Commands>
  a: Add files to archive
  b: Benchmark
  d: Delete files from archive
  e: Extract files from archive (without using directory names)
  l: List contents of archive
  t: Test integrity of archive
  u: Update files to archive
  x: eXtract files with full paths
<Switches>
  -ai[r[-|0]]{@listfile|!wildcard}: Include archives
  -ax[r[-|0]]{@listfile|!wildcard}: eXclude archives
  -bd: Disable percentage indicator
  -i[r[-|0]]{@listfile|!wildcard}: Include filenames
  -m{Parameters}: set compression Method
  -o{Directory}: set Output directory
  -p{Password}: set Password
  -r[-|0]: Recurse subdirectories
  -scs{UTF-8 | WIN | DOS}: set charset for list files
  -sfx[{name}]: Create SFX archive
  -si[{name}]: read data from stdin
  -slt: show technical information for l (List) command
  -so: write data to stdout
  -ssc[-]: set sensitive case mode
  -t{Type}: Set type of archive
  -u[-][p#][q#][r#][x#][y#][z#][!newArchiveName]: Update options
  -v{Size}[b|k|m|g]: Create volumes
  -w[{path}]: assign Work directory. Empty path means a temporary directory
  -x[r[-|0]]]{@listfile|!wildcard}: eXclude filenames
  -y: assume Yes on all queries
```

This does not seem to include any of the switches `-bso1`, `-bsp1` and `-bse2` !?!

**Comment by janusw** [comment]
On another machine I have a more recent Ubuntu version (19.04), which has p7zip version 16.02.

This version indeed seems to include the necessary switches:

```
  -bs{o|e|p}{0|1|2} : set output stream for output/error/progress line
```

So version 9.20 is apparently just too old! Indeed `make prepare` seems to have a check for p7zip, but it does not complain about version 9.20.

A remaining question is: Which minimum version of p7zip is required?

**Comment by janusw** [comment]
> A remaining question is: Which minimum version of p7zip is required?

https://github.com/jbdemonte/p7zip/blob/master/ChangeLog seems to indicate that the `-bs` switches were introduced with version 15.05. Any 9.x version is certainly too old.

**Comment by janusw** [comment]
I have verified that using p7zip version 16.02 (compiled from source) fixes the problem and makes `make prepare` run through without errors on Ubuntu 16.04.

It should be noted that the p7zip-full system package needs to stay installed, otherwise the checks of `make prepare` do not detect 7zip (even if a compiled-from-source 7za executable is available in the $PATH). Ideally those checks should detect the presence of a 7za executable instead of an installed deb package.

I can see two solutions for fixing this issue:
1. Fix the check to require version 15+ of p7zip or
2. Leave out the additional flags if an older version is detected.

**Comment by jonathanpeppers** [comment]
@grendello might have some insight here.

**Comment by janusw** [comment]
Thanks a lot for the quick fix! I can confirm that 07cc8cb8567cf608f23e08f7978dadbbb642f840 fixes the problem.

[Note: I still see errors in the actual build step on Ubuntu 16.04, but I will report those in a separate issue.]
